app/models/invoice.rb

```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class Invoice < ApplicationRecord
+  has_one :sale, dependent: :restrict_with_exception
```

Comment:
The place we use restrict_with_error is because when we try to delete the record we don't know if it will fail or not and if it fail we soft delete it. But here, it's not planned to be deleted in the code so it's with an exception.

Reply:
Got it! I think it's worth adding some comments on that, and some tests on the `destroy` code path (not only `soft_delete`). Would use `restrict_with_exception` everywhere and explicitly handle the exception and soft delete when it happens be more explicit? Not sure if it's very ruby-like though, I will defer to @nymous for this

Reply:
Thinking about it again, if it's expected to try to delete an article but we should block the deletion when it's currently used by a sale, having an error instead of an exception makes more sense

Reply:
Yes that has been raised by @nymous in one of his comment